### PR TITLE
Fix support for non-final classes in AtomicReference

### PR DESCRIPTION
**Pull request by lorentey**
Swift 5.6/7 generates new compiler warnings when non-final classes conform to `AtomicReference`. The warnings are pointing out a real design issue: Self is covariant in protocols, but we intend atomic operations to always deal with the base class rather than individual subclasses.

Add a new associated type requirement to AtomicValue to represent the actual type that originally conformed to the protocol, and add explicit same-type requirements to clients to ensure they aren’t asked to return a subclass.

Unfortunately, this change is source breaking:

```
class Base: AtomicReference {}
class Derived: Base {}

let ref = ManagedAtomic<Derived?>(nil)
  // before: builds, but only because a type safety issue isn't caught by the compiler
  // after: 'ManagedAtomic' requires the types 'Derived' and 'Base' be equivalent
```

### Checklist
- [X] I've read the [Contribution Guidelines](/README.md#contributing-to-swift-atomics)
- [X] My contributions are licensed under the [Swift license](/LICENSE.txt).
- [X] I've followed the coding style of the rest of the project.
- [ ] I've added tests covering all new code paths my change adds to the project (if appropriate).
- [ ] I've verified that my change does not break any existing tests.
- [ ] I've updated the documentation if necessary.

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
The source breakage seems mostly theoretical; it feels like the risk may be low enough to ship this in a minor release, especially since the compiler warning in 5.7 is highlighting a real correctness issue.

(If it proves otherwise, we might be able to do something about it, either by continuing to allow broken code in Swift 5.6, or by tweaking `UnsafeAtomic`/`ManagedAtomic` to allow taking subclasses as their generic argument. The latter would require the addition of ambiguous overloads marked `@_disfavoredOverload`, so I'd rather not do it if at all possible.)

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test macOS platform

**Comment by lorentey** [comment]
```
[7/22] Compiling Atomics AtomicOptionalRawRepresentable.swift
/build/swift-atomics/Sources/Atomics/AtomicOptionalRawRepresentable.swift:34:14: error: type of expression is ambiguous without more context
    _storage = Storage(value?.rawValue)
    ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
```

Oh, a fix for that is in #73.

**Comment by lorentey** [comment]
@swift-ci test

File: Sources/Atomics/AtomicStrongReference.swift [hunk 1]
```diff
@@ -13,17 +13,59 @@
...
+///     let value = ref.load(ordering: .relaxed)
+///     // `value` is a variable of type `MyObject`, holding a `Derived` instance.
+///     print(value is Derived) // Prints "true"
+///
 public protocol AtomicReference: AnyObject, AtomicOptionalWrappable
 where
-  AtomicRepresentation == AtomicReferenceStorage<Self>,
-  AtomicOptionalRepresentation == AtomicOptionalReferenceStorage<Self>
+  AtomicRepresentation == AtomicReferenceStorage<_AtomicBase>,
+  AtomicOptionalRepresentation == AtomicOptionalReferenceStorage<_AtomicBase>
```
Comment (winstondu) [inline Sources/Atomics/AtomicStrongReference.swift:49]: @lorentey Wow, this is a cool pattern. I did not know you could use a recursive type constraint within the where clause in the protocol inheritance itself! 

Any documentation on how this was available?

**Pull request merged**